cf/nbc.c

```diff
@@ -0,0 +1,145 @@
+ * Bugs guaranteed for failures in following this advice. */
+#define BYTES_PER_RECORD 4
+
+#define PRIOR_BAIS 0.5
+
+void bayes_learn(FILE *data_file, char category){
+    /* This function increases the counter, corresponding to the category, of hashes
+     * inputted from stdin in the specified data file.
+     * Arguments:
+     * data_file shall be a FILE pointer opened as "r+b".
+     * category shall be a char and equal to either 'T' or 'F'.
+     * Outcome:
+     * The corresponding counter will be increased. */
+    unsigned long hash;
+    long offset;
+    uint32_t counter, new_counter;
+
+    while (scanf("%lu", &hash) != -1){
+        offset = 2 * BYTES_PER_RECORD * (long)(hash % FOLD_TO);
+        offset += (category == 'T') ? 0 : (BYTES_PER_RECORD);
+
+        fseek(data_file, offset, SEEK_SET);
+        fread(&counter, BYTES_PER_RECORD, 1, data_file);
+
+        new_counter = counter + 1;
+        if (new_counter){
+            fseek(data_file, offset, SEEK_SET);  /* Go back. */
+            fwrite(&new_counter, BYTES_PER_RECORD, 1, data_file);
+        }
+        /* Else the original counter is already at maximum. Skip. */
+    }
+    return;
+}
+
+char bayes_classify(FILE *data_file){
+    /* This function performs Bayesian classification on the group of hashes inputted
+     * from stdin, using the learned data in the specified dat file.
+     * Arguments:
+     * data_file shall be a FILE pointer opened as "rb".
+     * Outcome:
+     * A char will be returned that is either 'T' or 'F', representing true or false
+     * positive. */
+    unsigned long hash;
+    long offset;
+
+    uint32_t tp_count, fp_count;
+    unsigned long long total_count;
+
+    long double weight;
+    long double p_hash_given_tp, p_hash_given_fp;
+    long double p_w_hash_given_tp, p_w_hash_given_fp;
+
+    long double prior = PRIOR_BAIS, posterior;
+
+    while (scanf("%lu", &hash) != -1){
+        offset = 2 * BYTES_PER_RECORD * (long)(hash % FOLD_TO);
+
+        fseek(data_file, offset, SEEK_SET);
+        fread(&tp_count, BYTES_PER_RECORD, 1, data_file);
+        fread(&fp_count, BYTES_PER_RECORD, 1, data_file);
+
+        total_count = (unsigned long long)(tp_count) + (unsigned long long)(fp_count);
+        if (total_count){
+            weight = logistic(total_count);
+
+            /* Seperate calculation needed as floating point accuracy might not be enough. */
+            p_hash_given_tp = (long double)(tp_count) / total_count;
+            p_hash_given_fp = (long double)(fp_count) / total_count;
+
+            p_w_hash_given_tp = (p_hash_given_tp - (long double)(0.5)) * weight + (long double)(0.5);
+            p_w_hash_given_fp = (p_hash_given_fp - (long double)(0.5)) * weight + (long double)(0.5);
+
+            /* Bayes theorem */
+            posterior = p_w_hash_given_tp * prior / (p_w_hash_given_tp * prior + p_w_hash_given_fp * ((long double)(1) - prior));
+            prior = posterior;
```

Comment:
The above two lines can be concatenated by `prior = p_w_hash_given_tp * prior / (p_w_hash_given_tp * prior + p_w_hash_given_fp * ((long double)(1) - prior));`.